screwdriver/screwdriver.go

```diff
@@ -0,0 +1,200 @@
+type Step struct {
+	Name    string `json:"name"`
+	Command string `json:"command"`
+}
+
+// Job is job entity struct
+type Job struct {
+	Steps       []Step            `json:"commands"`
+	Environment map[string]string `json:"environment"`
+	Image       string            `json:"image"`
+}
+
+type jobs map[string][]Job
+
+type validatorResponse struct {
+	Jobs   jobs     `json:"jobs"`
+	Errors []string `json:"errors"`
+}
+
+type tokenResponse struct {
+	JWT string `json:"token"`
+}
+
+// New creates a API
+func New(apiURL, token string) (API, error) {
+	s := &sdAPI{
+		HTTPClient: http.DefaultClient,
+		APIURL:     apiURL,
+		UserToken:  token,
+	}
+
+	jwt, err := s.jwt()
+	if err != nil {
+		return s, err
+	}
+
+	s.SDJWT = jwt
+
+	return s, nil
+}
+
+func (sd *sdAPI) makeURL(endpoint string) (*url.URL, error) {
+	u, err := url.Parse(sd.APIURL)
+	if err != nil {
+		return nil, err
+	}
+	u.Path = path.Join(u.Path, apiVersion, endpoint)
+
+	return u, nil
+}
+
+func (sd *sdAPI) request(method, path string, body io.Reader) (*http.Response, error) {
+	req, err := http.NewRequest(method, path, body)
+	if err != nil {
+		return nil, err
+	}
+
+	switch method {
+	case http.MethodGet:
+		{
+			req.Header.Add("Accept", "application/json")
+		}
+	case http.MethodPost, http.MethodPut, http.MethodDelete:
+		{
+			req.Header.Add("Content-Type", "application/json")
+			req.Header.Add("Authorization", "Bearer "+sd.SDJWT)
+		}
+	}
+
+	return sd.HTTPClient.Do(req)
+}
+
+func (sd *sdAPI) jwt() (string, error) {
+	fullpath, err := sd.makeURL(tokenEndpoint)
+	if err != nil {
+		return "", fmt.Errorf("failed to make request url: %v", err)
+	}
+
+	query := fullpath.Query()
+	query.Set("api_token", sd.UserToken)
+	fullpath.RawQuery = query.Encode()
+
+	res, err := sd.request(http.MethodGet, fullpath.String(), nil)
+	if err != nil {
+		return "", fmt.Errorf("failed to send request: %v", err)
+	}
+	if res.StatusCode != http.StatusOK {
+		return "", fmt.Errorf("failed to get JWT: StatusCode %d", res.StatusCode)
+	}
+	defer res.Body.Close()
+
+	tokenResponse := new(tokenResponse)
+	err = json.NewDecoder(res.Body).Decode(tokenResponse)
+	if err != nil {
+		return "", fmt.Errorf("failed to parse JWT response: %v", err)
+	}
+
+	return tokenResponse.JWT, err
+}
+
+func readScrewdriverYAML(filePath string) (string, error) {
+	yaml, err := ioutil.ReadFile(filePath)
+	if err != nil {
+		return "", fmt.Errorf("failed to read screwdriver.yaml: %v", err)
+	}
+	return string(yaml), nil
+}
+
+func (sd *sdAPI) validate(filePath string) (*validatorResponse, error) {
+	fullpath, err := sd.makeURL(validatorEndpoint)
+	if err != nil {
+		return &validatorResponse{}, fmt.Errorf("failed to make request url: %v", err)
+	}
+
+	yaml, err := readScrewdriverYAML(filePath)
+	if err != nil {
+		return &validatorResponse{}, err
+	}
+
+	escapedYaml := strings.ReplaceAll(string(yaml), "\"", "\\\"")
+	body := fmt.Sprintf(`{"yaml": "%s"}`, strings.ReplaceAll(string(escapedYaml), "\n", "\\n"))
```

Comment:
It looks string cast is not necessarily, because ReplaceAll return string.
https://golang.org/pkg/strings/#ReplaceAll

Reply:
Fixed